File: common/telegram.py
# coding: utf-8

import json
import urllib.parse
import urllib.request


class TelegramClient(object):
    def __init__(self, bot_token, api_base='https://api.telegram.org', timeout=20):
        self.bot_token = str(bot_token or '').strip()
        self.api_base = str(api_base or 'https://api.telegram.org').rstrip('/')
        self.timeout = int(timeout or 20)
        if not self.bot_token:
            raise ValueError('缺少 Telegram Bot Token')

    def api_url(self, method):
        return '{}/bot{}/{}'.format(self.api_base, self.bot_token, method)

    def request(self, method, data=None, timeout=None):
        payload = None
        if data is not None:
            payload = urllib.parse.urlencode(data).encode('utf-8')
        request = urllib.request.Request(self.api_url(method), data=payload, method='POST' if payload else 'GET')
        request_timeout = int(timeout or self.timeout)
        with urllib.request.urlopen(request, timeout=request_timeout) as response:
            body = response.read().decode('utf-8')
        result = json.loads(body)
        if not result.get('ok'):
            raise RuntimeError('Telegram API 返回失败：{}'.format(body))
        return result.get('result')

    def get_updates(self, offset=None, timeout=25):
        poll_timeout = int(timeout)
        data = {'timeout': poll_timeout}
        if offset:
            data['offset'] = int(offset)
        return self.request('getUpdates', data, timeout=poll_timeout + 5)

    def send_message(self, chat_id, text, parse_mode='HTML', disable_notification=False, reply_markup=None):
        data = {
            'chat_id': str(chat_id),
            'text': text,
            'parse_mode': parse_mode,
            'disable_web_page_preview': 'true',
            'disable_notification': 'true' if disable_notification else 'false'
        }
        if reply_markup:
            data['reply_markup'] = json.dumps(reply_markup, ensure_ascii=False)
        return self.request('sendMessage', data)

    def answer_callback_query(self, callback_query_id, text='', show_alert=False):
        data = {
            'callback_query_id': str(callback_query_id),
            'show_alert': 'true' if show_alert else 'false'
        }
        if text:
            data['text'] = text
        return self.request('answerCallbackQuery', data)

    def edit_message_text(self, chat_id, message_id, text, parse_mode='HTML', reply_markup=None):
        data = {
            'chat_id': str(chat_id),
            'message_id': str(message_id),
            'text': text,
            'parse_mode': parse_mode,
            'disable_web_page_preview': 'true'
        }
        if reply_markup:
            data['reply_markup'] = json.dumps(reply_markup, ensure_ascii=False)
        return self.request('editMessageText', data)

File: cf-worker/src/telegram.ts
// Telegram Bot API Client for Cloudflare Workers

export interface InlineKeyboardMarkup {
  inline_keyboard: Array<Array<{ text: string; callback_data: string }>>;
}

export interface TelegramUpdate {
  update_id: number;
  message?: TelegramMessage;
  edited_message?: TelegramMessage;
  callback_query?: CallbackQuery;
}

export interface TelegramMessage {
  message_id: number;
  from?: { id: number; is_bot: boolean; first_name: string; username?: string };
  chat: { id: number; type: string };
  date: number;
  text?: string;
}

export interface CallbackQuery {
  id: string;
  from: { id: number };
  message?: TelegramMessage;
  data?: string;
}

export class TelegramClient {
  private botToken: string;
  private apiBase: string;

  constructor(botToken: string, apiBase = 'https://api.telegram.org') {
    this.botToken = botToken;
    this.apiBase = apiBase.replace(/\/$/, '');
  }

  private apiUrl(method: string): string {
    return `${this.apiBase}/bot${this.botToken}/${method}`;
  }

  private async request<T = any>(method: string, data?: Record<string, any>): Promise<T> {
    const options: RequestInit = { method: 'GET' };
    if (data !== undefined) {
      options.method = 'POST';
      options.headers = { 'Content-Type': 'application/json' };
      options.body = JSON.stringify(data);
    }
    const response = await fetch(this.apiUrl(method), options);
    const result = await response.json() as any;
    if (!result.ok) {
      throw new Error(`Telegram API error: ${JSON.stringify(result)}`);
    }
    return result.result;
  }

  async sendMessage(
    chatId: string | number,
    text: string,
    replyMarkup?: InlineKeyboardMarkup
  ): Promise<any> {
    const data: Record<string, any> = {
      chat_id: String(chatId),
      text,
      parse_mode: 'HTML',
      disable_web_page_preview: true,
    };
    if (replyMarkup) {
      data.reply_markup = replyMarkup;
    }
    return this.request('sendMessage', data);
  }

  async answerCallbackQuery(callbackQueryId: string, text = '', showAlert = false): Promise<any> {
    return this.request('answerCallbackQuery', {
      callback_query_id: callbackQueryId,
      text,
      show_alert: showAlert,
    });
  }

  async editMessageText(
    chatId: string | number,
    messageId: number,
    text: string,
    replyMarkup?: InlineKeyboardMarkup
  ): Promise<any> {
    const data: Record<string, any> = {
      chat_id: String(chatId),
      message_id: String(messageId),
      text,
      parse_mode: 'HTML',
      disable_web_page_preview: true,
    };
    if (replyMarkup) {
      data.reply_markup = replyMarkup;
    }
    return this.request('editMessageText', data);
  }

  async setWebhook(url: string, secretToken?: string): Promise<any> {
    const data: Record<string, any> = { url };
    if (secretToken) {
      data.secret_token = secretToken;
    }
    return this.request('setWebhook', data);
  }

  async deleteWebhook(): Promise<any> {
    return this.request('deleteWebhook', { drop_pending_updates: true });
  }

  async getUpdates(offset?: number | null, timeout = 25): Promise<TelegramUpdate[]> {
    const data: Record<string, any> = { timeout };
    if (offset) data.offset = offset;
    return this.request('getUpdates', data);
  }
}
